File: experiments/complexity_analysis/On3_cubic/reference_solutions/solution.py
def matrix_multiply(A, B):
    """O(n³) cubic time: standard matrix multiplication algorithm"""
    n = len(A)
    C = [[0 for _ in range(n)] for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]
    
    return C

# Read input
n = int(input())

# Read matrix A
A = []
for i in range(n):
    row = list(map(int, input().split()))
    A.append(row)

# Read matrix B
B = []
for i in range(n):
    row = list(map(int, input().split()))
    B.append(row)

# Calculate C = A × B using optimal O(n³) approach
C = matrix_multiply(A, B)

# Calculate sum of all elements in result matrix
matrix_sum = sum(sum(row) for row in C)

# Output sum
print(matrix_sum)

File: experiments/complexity_analysis/On3_cubic/reference_solutions/solution.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<long long>> matrixMultiply(const vector<vector<long long>>& A, const vector<vector<long long>>& B) {
    // O(n³) cubic time: standard matrix multiplication algorithm
    int n = A.size();
    vector<vector<long long>> C(n, vector<long long>(n, 0));
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return C;
}

int main() {
    int n;
    cin >> n;
    
    // Read matrix A
    vector<vector<long long>> A(n, vector<long long>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> A[i][j];
        }
    }
    
    // Read matrix B
    vector<vector<long long>> B(n, vector<long long>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> B[i][j];
        }
    }
    
    // Calculate C = A × B using optimal O(n³) approach
    vector<vector<long long>> C = matrixMultiply(A, B);
    
    // Calculate sum of all elements in result matrix
    long long sum = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            sum += C[i][j];
        }
    }
    
    // Output sum
    cout << sum << endl;
    
    return 0;
}
